plugins/jspsych-video.js

```diff
@@ -97,20 +97,33 @@ jsPsych.plugins.video = (function() {
       video_html += trial.prompt;
     }
 
-    display_element.innerHTML = video_html;
+    var video_preload_blob = jsPsych.pluginAPI.getVideoBuffer(file_name);
+    if(!video_preload_blob) {
```

Comment:
Do you think it is possible/desirable to move this conditional inside the getVideoBuffer function? If the function could return something that worked for preloaded and non-preloaded video that could simplify the implementation requirements for plugins that add video.

Reply:
To do everything within getVideoBuffer, we'd need to pass getVideoBuffer the display Element and a partly formed "<video " node. 

But I can simplify it (see below). 

var file_name = trial.sources[0];

        var video_preload_blob = jsPsych.pluginAPI.getVideoBuffer(file_name);
        if(!video_preload_blob) {
            if(file_name.indexOf('?') > -1){
                file_name = file_name.substring(0, file_name.indexOf('?'));
            }
            var type = file_name.substr(file_name.lastIndexOf('.') + 1);
            type = type.toLowerCase();
            video_html+='><source src="' + file_name + '" type="video/'+type+'"></video>';
            display_element.innerHTML = video_html;
        }
        else{
            video_html+=">"
            display_element.innerHTML = video_html;
            display_element.childNodes[0].src = video_preload_blob;
        }

        display_element.querySelector('#jspsych-video-player').onended = function(){
            end_trial();
        }

        // event handler to set timeout to end trial if video is stopped
        display_element.querySelector('#jspsych-video-player').onplay = function(){
            if(trial.stop !== null){
                if(trial.start == null){
                    trial.start = 0;
                }
                jsPsych.pluginAPI.setTimeout(end_trial, (trial.stop-trial.start)*1000);
            }
        }

Reply:
any thought Josh?

Reply:
I'm struggling a little to understand exactly what's happening, so apologies if this doesn't make sense.

Since you set the `src` of the video object to be the output of the call to `getVideoBuffer`, couldn't the function just return the appropriate path if the video was not preloaded? 

Maybe the answer is no because you need to use `<source>` tags. In this case could the `getVideoBuffer` function actually return the whole video player object? Maybe by creating a [document fragment](https://developer.mozilla.org/en-US/docs/Web/API/Document/createDocumentFragment) that can be appended to the DOM inside the plugin?

Reply:
But I did it that way because you created the video originally by creating a string specifying the node. 

getVideoBuffer returns undefined when no video is prebuffered as getAudioBuffer does it this way. 

I worry that Document Fragment usage will lead to backward compatibilty issues over webbrowsers.

 var video_html = '<video id="jspsych-video-player"'
        if(trial.width) {
            video_html += ' width="'+trial.width+'"'
        }
        if(trial.height) {
            video_html += ' height="'+trial.height+'"'
        }
        if(trial.autoplay){
            video_html += " autoplay "
        }
        if(trial.controls){
            video_html +=" controls "
        }
        //show prompt if there is one
        if (trial.prompt !== null) {
            video_html += trial.prompt;
        }

        var file_name = trial.sources[0];

        var video_preload_blob = jsPsych.pluginAPI.getVideoBuffer(file_name);
        if(!video_preload_blob) {
            if(file_name.indexOf('?') > -1){
                file_name = file_name.substring(0, file_name.indexOf('?'));
            }
            var type = file_name.substr(file_name.lastIndexOf('.') + 1);
            type = type.toLowerCase();
            video_html+='><source src="' + file_name + '" type="video/'+type+'"></video>';
            display_element.innerHTML = video_html;
        }
        else{
            video_html+=">"
            display_element.innerHTML = video_html;
            display_element.childNodes[0].src = video_preload_blob;
        }